### ISSUE TITLE
Uri.fromString throws on some invalid port numbers

### ISSUE THREAD

**Issue by niij** [http4s/http4s#2583]
`org.http4s.Uri.fromString` throws when the port number is greater than 2,147,483,647  (`java.lang.Integer.MAX_VALUE`).

Versions
Present in 0.20.1 and 0.18.23

```scala
$ amm
Loading...
Welcome to the Ammonite Repl 1.1.2
(Scala 2.12.6 Java 1.8.0_181)
If you like Ammonite, please support our development at www.patreon.com/lihaoyi
@ import $ivy.`org.http4s::http4s-core:0.18.23`
import $ivy.$

@ org.http4s.Uri.fromString("https://example.com:2147483647")
res1: org.http4s.package.ParseResult[org.http4s.Uri] = Right(
  Uri(
    Some(Scheme(https)),
    Some(Authority(None, RegName(example.com), Some(2147483647))),
    "",
    Query(),
    None
  )
)

@ // Greater than java.lang.Integer.MAX_VALUE

@ org.http4s.Uri.fromString("https://example.com:2147483648")
java.lang.NumberFormatException: For input string: "2147483648"
  java.lang.NumberFormatException.forInputString(NumberFormatException.java:65)
  java.lang.Integer.parseInt(Integer.java:587)
  java.lang.Integer.parseInt(Integer.java:615)
  scala.collection.immutable.StringLike.toInt(StringLike.scala:301)
  scala.collection.immutable.StringLike.toInt$(StringLike.scala:301)
  scala.collection.immutable.StringOps.toInt(StringOps.scala:29)
  org.http4s.parser.Rfc3986Parser.Port(Rfc3986Parser.scala:85)
  org.http4s.parser.Rfc3986Parser.Port$(Rfc3986Parser.scala:83)
  org.http4s.parser.RequestUriParser.Port(RequestUriParser.scala:8)
  org.http4s.parser.Rfc3986Parser.Authority(Rfc3986Parser.scala:64)
  org.http4s.parser.Rfc3986Parser.Authority$(Rfc3986Parser.scala:63)
  org.http4s.parser.RequestUriParser.Authority(RequestUriParser.scala:8)
  org.http4s.parser.Rfc3986Parser.HierPart(Rfc3986Parser.scala:40)
  org.http4s.parser.Rfc3986Parser.HierPart$(Rfc3986Parser.scala:39)
  org.http4s.parser.RequestUriParser.HierPart(RequestUriParser.scala:8)
  org.http4s.parser.Rfc3986Parser.AbsoluteUri(Rfc3986Parser.scala:25)
  org.http4s.parser.Rfc3986Parser.AbsoluteUri$(Rfc3986Parser.scala:24)
  org.http4s.parser.RequestUriParser.AbsoluteUri(RequestUriParser.scala:8)
  org.http4s.parser.Rfc3986Parser.Uri(Rfc3986Parser.scala:22)
  org.http4s.parser.Rfc3986Parser.Uri$(Rfc3986Parser.scala:22)
  org.http4s.parser.RequestUriParser.Uri(RequestUriParser.scala:8)
  org.http4s.Uri$.$anonfun$fromString$1(Uri.scala:145)
  org.http4s.internal.parboiled2.Parser.runRule$1(Parser.scala:143)
  org.http4s.internal.parboiled2.Parser.phase0_initialRun$1(Parser.scala:151)
  org.http4s.internal.parboiled2.Parser.__run(Parser.scala:204)
  org.http4s.Uri$.fromString(Uri.scala:145)
  ammonite.$sess.cmd2$.<init>(cmd2.sc:1)
  ammonite.$sess.cmd2$.<clinit>(cmd2.sc)
```

**Comment by kevinmeredith** [comment]
Hey @niij - I'm not sure that I follow the problem. 

Based on the following, it's not possible to construct a `Uri.Authority

Example:

```scala
$amm
Loading...
Welcome to the Ammonite Repl 1.0.3
(Scala 2.12.4 Java 1.8.0_112)
If you like Ammonite, please support our development at www.patreon.com/lihaoyi
@ import $ivy.`org.http4s::http4s-core:0.18.23` 
...
import $ivy.$                                

@ org.http4s.Uri.Authority(port = Some(2147483648)) 
(console):1: integer number too large
org.http4s.Uri.Authority(port = Some(2147483648))

// but you can do something nonsensical (negative port numbers don't exist, do they?)
// like the following
@ org.http4s.Uri.Authority(port = Some(Int.MaxValue + 1)) 
res4: org.http4s.Uri.Authority = Authority(None, RegName(localhost), Some(-2147483648))
```

You're saying that `@ org.http4s.Uri.fromString("https://example.com:2147483648")` should've parsed successfully, i.e. using a port of `2147483648`?

**Comment by kevinmeredith** [comment]
Looking at https://github.com/http4s/http4s/blob/v0.18.23/core/src/main/scala/org/http4s/parser/Rfc3986Parser.scala#L83-L87, and then https://tools.ietf.org/html/rfc3986#page-22, I don't see any mention of an `Int` constraint.

```
3.2.3.  Port

   The port subcomponent of authority is designated by an optional port
   number in decimal following the host and delimited from it by a
   single colon (":") character.

      port        = *DIGIT

   A scheme may define a default port.  For example, the "http" scheme
   defines a default port of "80", corresponding to its reserved TCP
   port number.  The type of port designated by the port number (e.g.,
   TCP, UDP, SCTP) is defined by the URI scheme.  URI producers and
   normalizers should omit the port component and its ":" delimiter if
   port is empty or if its value would be the same as that of the
   scheme's default.
```

Could you please shed more light on this one, @bryce-anderson, since I believe you [wrote it](https://github.com/http4s/http4s/blame/42dbd60e942c6036f75dfebc2b12d3614e362b7f/core/src/main/scala/org/http4s/parser/Rfc3986Parser.scala#L65)?

**Comment by niij** [comment]
@kevinmeredith I'd expect a `org.http4s.package.ParseResult[org.http4s.Uri]` which is defined as `type ParseResult[+A] = Either[ParseFailure, A]`. Throwing is neither a `Left` or ` Right` 😢 

As far as whether very large port numbers should be a ParseFailure or a URI, I don't really have an opinion there.

**Comment by rossabaker** [comment]
The pedantic answer is that we should change it to a BigInt here, because the spec puts absolutely no limit.  A practical answer is that it should remain an `Int`, and be validated as non-negative, and return a `Left` if it's out of bounds.  Another possible answer is a newtype for `Port`, with validation on construction to prevent non-negative.  Even then, we'd have an open question on the max.

Whatever we do, what we're currently doing is certainly wrong.

**Comment by kevinmeredith** [comment]
Hi @rossabaker - could you please take a look at https://github.com/http4s/http4s/pull/2664? Thanks!

**Comment by rossabaker** [comment]
Fixed by #2687.
